src/main/java/co/cask/publisher/PubSubOutputFormat.java

```diff
@@ -0,0 +1,218 @@
+      .setTotalTimeout(Duration.ofSeconds(maxRetryTimeout))
+      .setInitialRpcTimeout(Duration.ofSeconds(1))
+      .setMaxRpcTimeout(Duration.ofSeconds(5))
+      .build();
+  }
+
+  private BatchingSettings getBatchingSettings(long countSize, long bytesThreshold, long delayThreshold) {
+    BatchingSettings.Builder batchingSettingsBuilder = BatchingSettings.newBuilder();
+    batchingSettingsBuilder.setElementCountThreshold(countSize);
+    batchingSettingsBuilder.setRequestByteThreshold(bytesThreshold);
+    batchingSettingsBuilder.setDelayThreshold(Duration.ofMillis(delayThreshold));
+    return batchingSettingsBuilder.build();
+  }
+
+  /**
+   * Writer publishes messages to PubSub using the passed Publisher,
+   * batching and retrying are performed by Publisher based on how it is configured, If publishing a message fails
+   * we maintain error count and we throw exception when this error count exceeds a threshold.
+   */
+  public class PubSubRecordWriter extends RecordWriter<NullWritable, Text> {
+    private final Publisher publisher;
+    private final AtomicLong failures;
+    private final AtomicReference<Throwable> error;
+    private final long errorThrehold;
+    private final Set<ApiFuture> futures;
+
+    public PubSubRecordWriter(Publisher publisher, long errorThreshold) throws IOException {
+      this.publisher = publisher;
+      this.error = new AtomicReference<>();
+      this.errorThrehold = errorThreshold;
+      this.failures = new AtomicLong(0);
+      this.futures = ConcurrentHashMap.newKeySet();
+    }
+
+    @Override
+    public void write(NullWritable key, Text value) throws IOException {
+      handleErrorIfAny();
+      ByteString data = ByteString.copyFromUtf8(value.toString());
+      PubsubMessage message = PubsubMessage.newBuilder().setData(data).build();
+      ApiFuture future = publisher.publish(message);
+      futures.add(future);
+      ApiFutures.addCallback(future, new ApiFutureCallback<String>() {
+        @Override
+        public void onFailure(Throwable throwable) {
+          error.set(throwable);
+          failures.incrementAndGet();
+          futures.remove(future);
+        }
+
+        @Override
+        public void onSuccess(String s) {
+          futures.remove(future);
+        }
+      });
+    }
+
+    private void handleErrorIfAny() throws IOException {
+      if (failures.get() > errorThrehold) {
+        throw new IOException(
+          String.format("Publishing to PubSub failed for %s records due to %s", failures.get(), error.get()));
+      }
+    }
+
+    @Override
+    public void close(TaskAttemptContext taskAttemptContext) throws IOException {
+      try {
+        handleErrorIfAny();
+      } finally {
+        try {
+          // wait till the pending futures are complete before calling shutdown
+          while (!futures.isEmpty()) {
+            TimeUnit.MILLISECONDS.sleep(10);
+          }
+          publisher.shutdown();
```

Comment:
from what I remember about the API, you should be calling this first, as it will try to publish any remaining messages. Without this, isn't it possible to wait forever?

Reply:
`shutdown` will publish any remaining messages (batch) that are not published yet and wait till they are processed but we wouldn't get the status of those messages (success/failure), however when we are waiting on the futures ourselves, its guaranteed that the messages will be published as the delay threshold will enforce the batch to be published even if the size and count are less than threshold and we will be able to verify the status of all published messages.